src/Components/Shared/src/WebEventData.cs

```diff
@@ -53,36 +56,140 @@ private WebEventData(int browserRendererId, ulong eventHandlerId, EventFieldInfo
 
         public EventArgs EventArgs { get; }
 
-        private static EventArgs ParseEventArgsJson(ulong eventHandlerId, string eventArgsType, string eventArgsJson)
+        private static EventArgs ParseEventArgsJson(Renderer renderer, ulong eventHandlerId, string eventName, string eventArgsJson)
         {
             try
             {
-                return eventArgsType switch
+                if (TryGetStandardWebEventArgsType(eventName, out var eventArgsType))
                 {
-                    "change" => DeserializeChangeEventArgs(eventArgsJson),
-                    "clipboard" => Deserialize<ClipboardEventArgs>(eventArgsJson),
-                    "drag" => Deserialize<DragEventArgs>(eventArgsJson),
-                    "error" => Deserialize<ErrorEventArgs>(eventArgsJson),
-                    "focus" => Deserialize<FocusEventArgs>(eventArgsJson),
-                    "keyboard" => Deserialize<KeyboardEventArgs>(eventArgsJson),
-                    "mouse" => Deserialize<MouseEventArgs>(eventArgsJson),
-                    "pointer" => Deserialize<PointerEventArgs>(eventArgsJson),
-                    "progress" => Deserialize<ProgressEventArgs>(eventArgsJson),
-                    "touch" => Deserialize<TouchEventArgs>(eventArgsJson),
-                    "unknown" => EventArgs.Empty,
-                    "wheel" => Deserialize<WheelEventArgs>(eventArgsJson),
-                    "toggle" => Deserialize<EventArgs>(eventArgsJson),
-                    _ => throw new InvalidOperationException($"Unsupported event type '{eventArgsType}'. EventId: '{eventHandlerId}'."),
-                };
+                    // Special case for ChangeEventArgs because its value type can be one of
+                    // several types, and System.Text.Json doesn't pick types dynamically
+                    if (eventArgsType == typeof(ChangeEventArgs))
+                    {
+                        return DeserializeChangeEventArgs(eventArgsJson);
+                    }
+                }
+                else
+                {
+                    // For custom events, the args type is determined from the associated delegate
+                    eventArgsType = renderer.GetEventArgsType(eventHandlerId);
+                }
```

Comment:
The logic above is where we've changed from the old "category"-based type selection to the newer rules, i.e., for built-in known event names we have hardcoded logic, and for unknown event names we get the type from the delegate.

The reason we have to have the hardcoded logic for known event names (and can't just use the delegate in all cases) is for back-compat. People might set up an event handler delegate that receives `EventArgs` and expect that, at runtime, they receive whichever subclass corresponds to the actual event. We retain that edge-case feature for built-in event names, but for custom events you only receive what you declare.

Also it's quite nice that the built-in events don't have to even look at the delegate and remain closer to their historical implementation. Less chance of any obscure behavioral changes.

Reply:
Is the expectation that you author a custom JsonConverter if you needed to handle complex deserialization?

Reply:
Yes, same as with a JS interop call.